lib/blockchain.py

```diff
@@ -146,71 +201,152 @@ def size(self):
+                else:
+                    prb = (self.fork_byte_offset - checkpoint_size) // NetworkConstants.HEADER_SIZE_LEGACY
+                    pob = (full_size - self.fork_byte_offset) // NetworkConstants.HEADER_SIZE
+                    self._size = prb + pob
+        else:
+            self._size = 0
 
     def verify_header(self, header, prev_hash, target):
-        _hash = hash_header(header)
+        block_height = header.get('block_height')
+        _hash = hash_header(header, block_height)
+
         if prev_hash != header.get('prev_block_hash'):
             raise BaseException("prev hash mismatch: %s vs %s" % (prev_hash, header.get('prev_block_hash')))
-        if bitcoin.NetworkConstants.TESTNET:
-            return
         bits = self.target_to_bits(target)
-        if bits != header.get('bits'):
-            raise BaseException("bits mismatch: %s vs %s" % (bits, header.get('bits')))
+        if bits != int(header.get('bits'), 16):
+            raise BaseException("bits mismatch: %s vs %s" % (bits, int(header.get('bits'), 16)))
         if int('0x' + _hash, 16) > target:
             raise BaseException("insufficient proof of work: %s vs target %s" % (int('0x' + _hash, 16), target))
+        if is_postfork(block_height):
+            header_bytes = bytes.fromhex(serialize_header(header, False))
+            nonce = uint256_from_bytes(bfh(header.get('nonce'))[::-1])
+            solution = bfh(header.get('solution'))[::-1]
+            offset, length = var_int_read(solution, 0)
+            solution = solution[offset:]
+
+            if not is_gbp_valid(header_bytes, nonce, solution, NetworkConstants.EQUIHASH_N, NetworkConstants.EQUIHASH_K):
+                raise BaseException("Invalid equihash solution")
+
+    def verify_chunk(self, height, data):
+        size = len(data)
+        offset = 0
+        prev_hash = self.get_hash(height-1)
+
+        ser_chunks = ''
+        headers = {}
+        target = 0
+
+        while offset < size:
+            header_size = get_header_size(height)
+            raw_header = data[offset:(offset + header_size)]
+            header = deserialize_header(raw_header, height)
+
+            # Check retarget
+            if needs_retarget(height) or target == 0:
+                target = self.get_target(height, headers)
 
-    def verify_chunk(self, index, data):
-        num = len(data) // 80
-        prev_hash = self.get_hash(index * 2016 - 1)
-        target = self.get_target(index-1)
-        for i in range(num):
-            raw_header = data[i*80:(i+1) * 80]
-            header = deserialize_header(raw_header, index*2016 + i)
             self.verify_header(header, prev_hash, target)
-            prev_hash = hash_header(header)
+
+            headers[height] = header
+            prev_hash = hash_header(header, height)
+            offset += header_size
+            height += 1
+            ser_chunks += serialize_header(header)
```

Comment:
What's the difference between the original chunk and the re-serialized chunk?

Reply:
Can be optimized away. Will change it.

Reply:
Done